controllers/pipelinegates/pipelinegate_controller.go

```diff
@@ -0,0 +1,376 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinegates
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/google/uuid"
+
+	consoleclient "github.com/pluralsh/deployment-operator/pkg/client"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	client "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/go-logr/logr"
+	console "github.com/pluralsh/console-client-go"
+	pipelinesv1alpha1 "github.com/pluralsh/deployment-operator/api/pipelines/v1alpha1"
+
+	//job "k8s.io/api/batch/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PipelineGateReconciler reconciles a PipelineGate object
+type PipelineGateReconciler struct {
+	client.Client
+	ConsoleClient consoleclient.Client
+	Scheme        *runtime.Scheme
+	Log           logr.Logger
+}
+
+//+kubebuilder:rbac:groups=pipelines.plural.sh,resources=pipelinegates,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=pipelines.plural.sh,resources=pipelinegates/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=pipelines.plural.sh,resources=pipelinegates/finalizers,verbs=update
+//+kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete;deletecollection
+
+func (r *PipelineGateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	log := r.Log.WithValues("PipelineGate", req.NamespacedName)
+
+	gate := &pipelinesv1alpha1.PipelineGate{}
+
+	// get pipelinegate
+	if err := r.Get(ctx, req.NamespacedName, gate); err != nil {
+		if apierrs.IsNotFound(err) {
+			log.V(1).Info("PipelineGate CR not found - skipping.", "Namespace", gate.Namespace, "Name", gate.Name)
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Unable to fetch PipelineGate.")
+		return ctrl.Result{}, err
+	}
+
+	// INITIAL STATE
+	// on initial reconcile we set the state to pending
+	if gate.Status.State == nil {
+		// update CR state
+		gate.Status.State = &gate.Spec.SyncedState
+		if err := r.Status().Update(ctx, gate); err != nil {
+			log.Error(err, "Failed to update PipelineGate status at CR")
+			return ctrl.Result{}, err
+		}
+		log.V(1).Info("Updated state of CR on first reconcile.", "Namespace", gate.Namespace, "Name", gate.Name, "ID", gate.Spec.ID, "SyncedState", gate.Spec.SyncedState, "LastSyncedAt", gate.Spec.LastSyncedAt, "State", gate.Status.State)
+		return ctrl.Result{}, nil
+	}
+
+	// PENDING
+	if gate.Status.State != nil && *gate.Status.State == pipelinesv1alpha1.GateState(console.GateStatePending) {
```

Comment:
It would be nice to add methods on gate state to check that. Then conditions here would look simpler.

Reply:
I.e. like [here](https://github.com/pluralsh/console/blob/0ac7b9a976cb2aea159037047e9e931f32044dc1/controller/api/v1alpha1/cluster_types.go#L389).